### ISSUE TITLE
How to disable Micrometer metrics (with SpringBoot Actuator) and JMX auto-collection

### ISSUE THREAD

**Issue by adamisvk** [microsoft/ApplicationInsights-Java#1445]
Hi,
in SB Actuator, there are 2 ways to expose micrometer metrics per docs - Web (HTTP) or JMX.
https://docs.spring.io/spring-boot/docs/current/reference/html/production-ready-features.html#production-ready-endpoints

In docs for Codeless monitoring, it is said, that Micrometer and JMX Metrics are auto-collected and in configuration, you can disable collection of micrometer metrics.
https://docs.microsoft.com/en-us/azure/azure-monitor/app/java-standalone-config

I did disable collection of micrometer metrics, however they still seem to be collected (version 3.0). 
```
"micrometer": {
    "enabled": false
}
```
I suspect it is because Actuator by default exposes them via JMX too (in my config, I expose them via Web and JMX).

If that is true, my question is - is it possible to disable auto-collection of micrometer metrics exposed via JMX? Or possibly disable all auto-collection of JMX metrics? And maybe this distinction of exposure of actuator metrics could be mentioned in docs?

I am going to disable exposure of micrometer metrics by JMX in actuator config, I just wanted to know if it was possible to disable auto-collection of JMX metrics.

Thank you.

**Comment by ashwajce** [comment]
+1

**Comment by trask** [comment]
Hi @adamisvk @ashwajce!

(@adamisvk sorry for the late reply)

There was an issue with this configuration setting prior to 3.0.2, can you re-try with 3.0.2?

Thanks,
Trask

**Comment by LumirH** [comment]
We have the same problem on 3.0.2 app insights. 
I have disabled micrometer -- using following config:

![image](https://user-images.githubusercontent.com/31672125/107191004-da8f7d80-69eb-11eb-8d4e-6dcf169146d9.png)

Despite of that, I'm receiving lot of custom metrics in one minute interval:
![image](https://user-images.githubusercontent.com/31672125/107191175-1591b100-69ec-11eb-8713-cc63f9a789a3.png)

I'm not sure how are these metrics originated or what I should do to turn them off (or reduce the frequency).

**Comment by trask** [comment]
hi @LumirH, "GC Total Time" metric is part of the default metrics that the javaagent captures (it's not part of micrometer instrumentation), see https://github.com/microsoft/ApplicationInsights-Java/wiki/Default-Metrics-Captured-by-Java-3.0-Agent#default-metrics-captured-as-custom-metrics. We don't currently have a way to disable these default metrics.

**Comment by LumirH** [comment]
Hi @trask, thank you for the explanation. 
I'm actually getting more custom metrics, not just those listed on that page. 
For example:

- logback_events
- Suspected Deadlocked Threads
- tomcat_sessions_active_max
- tomcat_sessions_expired
- system_cpu_count

And others. Can I somehow control those?

My challenge is that we run application insights on the edge DMZ. 
We have limited network bandwidth available between these DMZs and application insights collector. 
These custom metrics are responsible for ~60% of the ingested data from those special microservices. 

It would be great if we can control the period how often are these data sent.

**Comment by trask** [comment]
Some of those metrics you listed above are part of micrometer, so it sounds like disabling micrometer instrumentation isn't working...

[a bit later]

oh, ya, it's only disabling part of the micrometer metrics instrumentation 🙁, I will fix shortly, those do produce a lot of metrics

**Comment by adamisvk** [comment]
Thank you very much! Btw, for anyone looking, I managed to at least lower the volume of custom metrics with following SB configuration (however, still receiving jvm, hikari, tomcat and other metrics... not really sure how, I tried to disable everything):

`management:
  metrics:
    enable:
      all: false
  endpoint:
    metrics:
      enabled: false
  endpoints:
    web:
      exposure:
        include: "*"
        exclude: "metrics"
    jmx:
      exposure:
        exclude: "*"`

**Comment by trask** [comment]
Here's the latest SNAPSHOT build from CI in case you have a chance to try it out: [applicationinsights-agent-3.0.3-BETA-SNAPSHOT.jar.zip](https://github.com/microsoft/ApplicationInsights-Java/files/5948975/applicationinsights-agent-3.0.3-BETA-SNAPSHOT.jar.zip)

**Comment by LumirH** [comment]
Hi @trask  -- I tested it and works well. 

Btw going through the logs I also found reportingInterval. It seems to be working as well.
This is exactly what I wanted (keep micrometer running, but in slow pace).

![image](https://user-images.githubusercontent.com/31672125/107360105-ccb62700-6ad5-11eb-8472-f16c2c3096ba.png)

I'll raise a new feature request for the "performance data" reporting interval.

**Comment by ashwajce** [comment]
@LumirH  @trask  Yes, With new 3.0.3 agent and reduced reporting interval, now appinsights collecting less custom metrics. Do you what is the default value of reporting interval?
New graph looks like this 
![image](https://user-images.githubusercontent.com/28860805/107531607-4e34b480-6bbd-11eb-8d52-c12c82e1711d.png)

**Comment by trask** [comment]
hey @LumirH @ashwajce (and everyone reading this issue)

the micrometer `reportingInternalSeconds` was an undocumented feature that we didn't hide well enough (and cleverly found!)

we added official (preview) support for this feature under a new setting in [3.0.3-BETA](https://github.com/microsoft/ApplicationInsights-Java/releases/tag/3.0.3-BETA) (and removed the old setting):

```
{
  "preview": {
    "metricIntervalSeconds": 60
  }
}
```

The new setting applies to all of these metrics:

* Default performance counters, e.g. CPU and Memory
* Default custom metrics, e.g. Garbage collection timing
* Configured JMX metrics
* Micrometer metrics

I just merged a doc PR with this new setting, so it should show up on https://docs.microsoft.com/en-us/azure/azure-monitor/app/java-standalone-config within the next 24 hours.
